src/Registration/cReg/tests/test_cReg.cpp

```diff
@@ -1144,6 +1146,48 @@ int main(int argc, char* argv[])
         std::cout << "//                  Finished weighted mean test.\n";
         std::cout << "//------------------------------------------------------------------------ //\n";
     }
+    {
+
+        std::cout << "// ----------------------------------------------------------------------- //\n";
+        std::cout << "//                  Starting CGP<->DVF test...\n";
+        std::cout << "//------------------------------------------------------------------------ //\n";
+
+        auto dvf_sptr = std::dynamic_pointer_cast<const NiftiImageData3DDeformation<float> >(
+                    NA.get_deformation_field_forward_sptr());
+
+        // DVF->CPG
+        float spacing[3];
+        for (unsigned i=0; i<3; ++i)
+            spacing[i] = dvf_sptr->get_raw_nifti_sptr()->pixdim[i+1] * 2.f;
+        NiftiImageData3DBSpline<float> dvf_to_cpg(*dvf_sptr, spacing);
+        if (std::abs(dvf_to_cpg.get_max()) < 1.e-4f || std::abs(dvf_to_cpg.get_min()) < 1.e-4f)
+            throw std::runtime_error("NiftiImageData3DBSpline::NiftiImageData3DBSpline(DVF): contains only zeroes.");
+
+        // DVF->CPG->DVF
+        auto dvf_to_cpg_to_dvf = dvf_to_cpg.get_as_deformation_field(*dvf_sptr->get_tensor_component(0));
+
+        NiftiImageData<float>::print_headers({ref_aladin.get(), dvf_sptr.get(),
+                                              &dvf_to_cpg, &dvf_to_cpg_to_dvf});
+
+        // Compare
+        if (*dvf_sptr != dvf_to_cpg_to_dvf)
+            throw std::runtime_error("DVF->CPG->DVF != DVF.");
+
+        // Do the same, using the converter
+        ControlPointGridToDeformationConverter<float> cpg_2_dvf_converter;
+        cpg_2_dvf_converter.set_cpg_spacing(spacing);
+        cpg_2_dvf_converter.set_reference_image(*dvf_sptr->get_tensor_component(0));
+        auto dvf_to_cpg_w_converter = cpg_2_dvf_converter.backward(*dvf_sptr);
+        auto dvf_to_cpg_to_dvf_w_converter = cpg_2_dvf_converter.forward(dvf_to_cpg_w_converter);
+
+        // Compare
+        if (dvf_to_cpg_to_dvf != dvf_to_cpg_to_dvf_w_converter)
+            throw std::runtime_error("ControlPointGridToDeformationConverter DVF->CPG->DVF failed.");
```

Comment:
Does this test work? It can't see why it should.In notation above, this is `BB' dvf`. Again, `BB'` shouldn't be identity.

Reply:
ah, bad error message. just checks that the result is the same whether the converter class is used or the methods inside of the `Deformation` and `BSpline` classes are called directly.

A couple of lines higher in the code, though, I do check that DVF->CPG->DVF does == DVF (checks `BB'` is identity). I agree that this doesn't have to be the case, but I suspected it would be very close to identity for a simple transformation, and where the CPG only has voxel sizes that are x2 the original DVF. It seemed like a good salt test to check the returned images weren't all zeroes, for example.

Reply:
i think that if the `normalise_kernel=true`, you'd be getting close to an inverse (but still not  100%).

anywa,y the important thing is the `adjoint` test.